codecparsers/bitReader.cpp

```diff
@@ -74,35 +81,31 @@ inline uint32_t BitReader::extractBitsFromCache(uint32_t nbits)
     return tmp;
 }
 
-uint32_t BitReader::read(uint32_t nbits)
+bool BitReader::read(uint32_t& v, uint32_t nbits)
 {
     assert(nbits <= (CACHEBYTES << 3));
 
-    /* Firstly loading data to m_cache, only need to read aligned bytes.
-       So we can load 8 bytes which aligned with machine address in the next per time*/
-    if (!m_loadBytes) {
-        uint32_t alignedBytes = CACHEBYTES - ((uintptr_t)m_stream % CACHEBYTES);
-        uint32_t toBeLoadSize = std::min(alignedBytes, m_size);
-        loadDataToCache(toBeLoadSize);
-    }
-
-    uint32_t res = 0;
-
     if (nbits <= m_bitsInCache) {
-        res = extractBitsFromCache(nbits);
-    } else { /*not enough bits, need to save remaining bits
-               in current cache and then reload more bits*/
-        uint32_t toBeReadBits = nbits - m_bitsInCache;
-        uint32_t tmp = extractBitsFromCache(m_bitsInCache);
-        uint32_t remainingBytes = m_size - m_loadBytes;
-        uint32_t toBeLoadSize = std::min(remainingBytes, CACHEBYTES);
-        toBeReadBits = std::min(toBeReadBits, toBeLoadSize << 3);
-        tmp <<= toBeReadBits;
-        loadDataToCache(toBeLoadSize);
-        res = tmp | extractBitsFromCache(toBeReadBits);
+        v = extractBitsFromCache(nbits);
+        return true;
     }
+    /*not enough bits, need to save remaining bits
+      in current cache and then reload more bits*/
+    uint32_t toBeReadBits = nbits - m_bitsInCache;
+    uint32_t tmp = extractBitsFromCache(m_bitsInCache);
+    reload();
+    if (toBeReadBits > m_bitsInCache)
+        return false;
```

Comment:
No, we need provide error information in higher level. Only higher level(codec parser in most of case) can print useful information. I will demo this in h264 parser. You can check it after send h264 fixes.